cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ApplicationSpecificationAdapter.java

```diff
@@ -165,6 +165,11 @@ private ApplicationSpecificationAdapter(SchemaGenerator schemaGenerator, Gson gs
         SortedMap.class.isAssignableFrom(rawType)) {
         return null;
       }
+
+      if (!ParameterizedType.class.isAssignableFrom(type.getType().getClass())) {
```

Comment:
I don't understand what you mean. Why it is just a HasMap without the type parameters? You can use `GSON.toJson(nodeStates, new TypeToken<Map<String, NodeState>>() { }.getType())`